IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -22,30 +22,218 @@
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Key is not valid. Make sure service and account are not nil.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Set keychain item with invalid key.");
+        return NO;
+    }
+
+    NSData *itemData = [serializer serialize:item];
+    if (!itemData)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to serialize token item.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Failed to serialize token item.");
+        return NO;
+    }
+    
+    NSMutableDictionary *query = [self defaultQuery];
+    [query setObject:key.service forKey:(id)kSecAttrService];
+    [query setObject:key.account forKey:(id)kSecAttrAccount];
+    // Backward compatibility with ADAL.
+    [query setObject:[s_libraryString dataUsingEncoding:NSUTF8StringEncoding] forKey:(id)kSecAttrGeneric];
+    // Backward compatibility with MSAL.
+    [query setObject:[NSNumber numberWithUnsignedInt:item.tokenType] forKey:(id)kSecAttrType];
+    
+    MSID_LOG_INFO(context, @"Trying to update keychain item...");
+    OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)@{(id)kSecValueData : itemData});
+    MSID_LOG_INFO(context, @"Keychain update status: %d", status);
+    if (status == errSecItemNotFound)
+    {
+        [query setObject:itemData forKey:(id)kSecValueData];
+        [query setObject:(id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly forKey:(id)kSecAttrAccessible];
+        
+        MSID_LOG_INFO(context, @"Trying to add keychain item...");
+        status = SecItemAdd((CFDictionaryRef)query, NULL);
+        MSID_LOG_INFO(context, @"Keychain add status: %d", status);
+    }
+    
+    if (status != errSecSuccess)
+    {
+        if (error)
+        {   
+            *error = MSIDCreateError(NSOSStatusErrorDomain, status, @"Failed to set item into keychain.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Failed to set item into keychain (status: %d)", status);
+    }
+    
+    return status == errSecSuccess;
 }
 
 - (MSIDToken *)itemWithKey:(MSIDTokenCacheKey *)key
                 serializer:(id<MSIDTokenSerializer>)serializer
                    context:(id<MSIDRequestContext>)context
                      error:(NSError **)error
 {
-    return nil;
+    MSID_LOG_INFO(context, @"itemWithKey:serializer:context:error:");
+    NSArray<MSIDToken *> *items = [self itemsWithKey:key serializer:serializer context:context error:error];
+    
+    if (items.count > 1)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorCacheMultipleUsers, @"The token cache store for this resource contains more than one user. Please set 'account' and 'service' properties of MSIDTokenCacheKey.", nil, nil, nil, context.correlationId, nil);
```

Comment:
nit: since this error will be probably propagated to the developer, we should perhaps modify the text to not have 
`Please set 'account' and 'service' properties of MSIDTokenCacheKey.`
because it's not actionable for the external developer

Reply:
done